src/System.Private.CoreLib/shared/System/Buffers/ArrayPool.cs

```diff
@@ -19,6 +19,10 @@ namespace System.Buffers
     /// </remarks>
     public abstract class ArrayPool<T>
     {
+        // Store the shared ArrayPool in a field of its derived sealed type so the Jit can "see" the exact type
+        // when the Shared property is inlined which will allow it to devirtualize calls made on it.
+        private readonly static TlsOverPerCoreLockedStacksArrayPool<T> s_shared = new TlsOverPerCoreLockedStacksArrayPool<T>();
```

Comment:
In the case of a reference type (or more generally where no boxing or conversion is needed), it seems like the C# compiler could be improved here slightly to work better with devirtualization to make the automatic backing field be of the type of the expression on the RHS rather than the type of the property, e.g.
```C#
public class C
{
    public static A Shared { get; } = new B();
}
public abstract class A { }
public sealed class B : A { }
```
produces
```C#
    private static readonly A <Shared>k__BackingField = new B();
    public static A Shared => <Shared>k__BackingField;
```
but it could instead do:
```C#
    private static readonly B <Shared>k__BackingField = new B();
    public static A Shared => <Shared>k__BackingField;
```
cc: @jaredpar

Reply:
I think that could only be done when there was an identity preserving conversion. This couldn't be done for instance when the conversion from B to A was boxing, user defined, etc ... Otherwise that change would be quite observable.

Reply:
It's Friday, my brain is mush. I didn't read your description carefully enough 😦

Reply:
Capturing link back to from proposal dotnet/roslyn#30797